xcdat/regridder/regrid2.py

```diff
@@ -46,259 +46,170 @@ def __init__(self, input_grid: xr.Dataset, output_grid: xr.Dataset, **options: A
-        ordered_axis_names : List[str]
-            List of axis name in order of dimensions of ``input_data``.
-
-        Returns
-        -------
-        np.ndarray
-            Multi-dimensional array on destination grid.
-        """
-        input_lat_index = ordered_axis_names.index("Y")
+    # convert to pure numpy
+    input_data = input_data_var.astype(np.float32).data
 
-        input_lon_index = ordered_axis_names.index("X")
+    y_name, y_index = _get_dimension(input_data_var, "Y")
+    x_name, x_index = _get_dimension(input_data_var, "X")
 
-        output_shape = [axis_sizes[x] for x in ordered_axis_names]
+    y_length = len(lat_mapping)
+    x_length = len(lon_mapping)
 
-        output_data = np.zeros(output_shape, dtype=np.float32)
+    other_dims = {
+        x: y for x, y in input_data_var.sizes.items() if x not in (y_name, x_name)
+    }
+    other_sizes = list(other_dims.values())
 
-        base_put_index = self._base_put_indexes(axis_sizes)
+    data_shape = [y_length * x_length] + other_sizes
+    output_data = np.zeros(data_shape, dtype=np.float32)
 
-        for lat_index, lat_map in enumerate(self._lat_mapping):
-            lat_weight = self._lat_weights[lat_index]
+    is_2d = input_data_var.ndim <= 2
 
-            input_lat_segment = np.take(input_data, lat_map, axis=input_lat_index)
+    # need to optimize
+    for y in range(y_length):
+        y_seg = np.take(input_data, lat_mapping[y], axis=y_index)
 
-            for lon_index, lon_map in enumerate(self._lon_mapping):
-                lon_weight = self._lon_weights[lon_index]
+        for x in range(x_length):
+            x_seg = np.take(y_seg, lon_mapping[x], axis=x_index, mode="wrap")
 
-                dot_weight = np.dot(lat_weight, lon_weight)
+            cell_weight = np.dot(lat_weights[y], lon_weights[x])
 
-                cell_weight = np.sum(dot_weight)
+            output_seg_index = y * x_length + x
 
-                input_lon_segment = np.take(
-                    input_lat_segment, lon_map, axis=input_lon_index
+            if is_2d:
+                output_data[output_seg_index] = np.divide(
+                    np.sum(
+                        np.multiply(x_seg, cell_weight),
+                        axis=(y_index, x_index),
+                    ),
+                    np.sum(cell_weight),
                 )
-
-                data = (
-                    np.nansum(
-                        np.multiply(input_lon_segment, dot_weight),
-                        axis=(input_lat_index, input_lon_index),
-                    )
-                    / cell_weight
+            else:
+                output_seg = output_data[output_seg_index]
+
+                np.divide(
+                    np.sum(
+                        np.multiply(x_seg, cell_weight),
+                        axis=(y_index, x_index),
+                    ),
+                    np.sum(cell_weight),
+                    out=output_seg,
```

Comment:
Comments explaining the logic here would be good. Maybe in the docstring.

Reply:
Resolved with https://github.com/xCDAT/xcdat/pull/533/commits/7695e2b89f16a430045c7e04f12c89fd1edbcdc6